Index: docker.io-app/engine/distribution/push.go
===================================================================
--- docker.io-app.orig/engine/distribution/push.go
+++ docker.io-app/engine/distribution/push.go
@@ -2,7 +2,6 @@ package distribution // import "github.c
 
 import (
 	"bufio"
-	"compress/gzip"
 	"context"
 	"fmt"
 	"io"
@@ -11,6 +10,7 @@ import (
 	"github.com/distribution/reference"
 	"github.com/docker/docker/api/types/events"
 	"github.com/docker/docker/pkg/progress"
+	"github.com/klauspost/compress/zstd"
 )
 
 const compressionBufSize = 32768
@@ -107,7 +107,11 @@ func compress(in io.Reader) (io.ReadClos
 	pipeReader, pipeWriter := io.Pipe()
 	// Use a bufio.Writer to avoid excessive chunking in HTTP request.
 	bufWriter := bufio.NewWriterSize(pipeWriter, compressionBufSize)
-	compressor := gzip.NewWriter(bufWriter)
+	compressor, err := zstd.NewWriter(bufWriter, zstd.WithEncoderLevel(1))
+	if err != nil {
+		pipeWriter.CloseWithError(err)
+		close(compressionDone)
+	}
 
 	go func() {
 		_, err := io.Copy(compressor, in)
